### PR TITLE
[ConstraintSystem] Avoid applying invalid solution with fixes

### PR DESCRIPTION
**Pull request by xedin**
If solution score indicates that there should be a fix but
none where recorded fail solution application and produce
a fallback diagnostic to pin-point a problem.

Couple of points:

* Start tracking number of holes involved in each solution, this allows us to:
  a. Score solutions based on number of holes as a fallback for "fix" weight;
  b. Identify cases where there are one or more holes but not "fixes".

* Detect misuse of `_` (discard assignment expression) while performing
  constraint generation which helps to avoid solutions with holes and no fixes
  which used to be diagnosed by syntactic diagnostics before (in `MiscDiagnostics`).

Resolves: rdar://problem/60663007

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Review by hborla — COMMENTED** [review]
Looks great! I just have one small comment about using `SK_Hole` in other places

File: lib/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -634,6 +634,8 @@ enum ScoreKind {
 
   /// A fix needs to be applied to the source.
   SK_Fix,
+  /// A hole in the constraint system.
+  SK_Hole,
```
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: Sure, I'll take a look!

[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: Done!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**